File: solutions/typescript/0030.ts
function findSubstring(s: string, words: string[]): number[] {
    if (!s || !words || words.length === 0) {
        return [];
    }
    
    const wordLen = words[0].length;
    const totalLen = words.length * wordLen;
    const result: number[] = [];
    
    // 创建words的计数映射
    const wordCount = new Map<string, number>();
    for (const word of words) {
        wordCount.set(word, (wordCount.get(word) || 0) + 1);
    }
    
    // 遍历所有可能的起始位置
    for (let i = 0; i <= s.length - totalLen; i++) {
        const currentCount = new Map<string, number>();
        let matched = 0;
        
        // 检查从位置i开始的子串
        for (let j = 0; j < words.length; j++) {
            const start = i + j * wordLen;
            const word = s.substring(start, start + wordLen);
            
            // 如果这个单词不在words中，直接跳出
            if (!wordCount.has(word)) {
                break;
            }
            
            // 更新当前计数
            currentCount.set(word, (currentCount.get(word) || 0) + 1);
            
            // 如果当前单词出现次数超过words中的次数，跳出
            if (currentCount.get(word)! > wordCount.get(word)!) {
                break;
            }
            
            matched++;
        }
        
        // 如果所有单词都匹配了，记录起始位置
        if (matched === words.length) {
            result.push(i);
        }
    }
    
    return result;
};

File: solutions/rust/0030.rs
use std::collections::HashMap;

impl Solution {
    pub fn find_substring(s: String, words: Vec<String>) -> Vec<i32> {
        if s.is_empty() || words.is_empty() {
            return vec![];
        }
        
        let word_len = words[0].len();
        let total_len = words.len() * word_len;
        let mut result = Vec::new();
        
        // 创建words的计数映射
        let mut word_count = HashMap::new();
        for word in &words {
            *word_count.entry(word.as_str()).or_insert(0) += 1;
        }
        
        // 使用滑动窗口优化：只需要检查 word_len 个不同的起始位置
        for start in 0..word_len {
            if start + total_len > s.len() {
                break;
            }
            
            let mut current_count = HashMap::new();
            let mut matched = 0;
            let mut left = start;
            
            // 初始化窗口
            for i in 0..words.len() {
                let word_start = start + i * word_len;
                let word_end = word_start + word_len;
                
                if word_end > s.len() {
                    break;
                }
                
                let word = &s[word_start..word_end];
                
                if word_count.contains_key(word) {
                    *current_count.entry(word).or_insert(0) += 1;
                    if current_count[word] <= word_count[word] {
                        matched += 1;
                    }
                }
            }
            
            if matched == words.len() {
                result.push(start as i32);
            }
            
            // 滑动窗口
            while left + total_len + word_len <= s.len() {
                // 移除左边的单词
                let left_word = &s[left..left + word_len];
                if word_count.contains_key(left_word) {
                    if current_count[left_word] <= word_count[left_word] {
                        matched -= 1;
                    }
                    *current_count.get_mut(left_word).unwrap() -= 1;
                }
                
                // 添加右边的单词
                let right_start = left + total_len;
                let right_word = &s[right_start..right_start + word_len];
                if word_count.contains_key(right_word) {
                    *current_count.entry(right_word).or_insert(0) += 1;
                    if current_count[right_word] <= word_count[right_word] {
                        matched += 1;
                    }
                }
                
                left += word_len;
                
                if matched == words.len() {
                    result.push(left as i32);
                }
            }
        }
        
        result
    }
}